### PR TITLE
Fix purecap byval arguments

### PR DESCRIPTION
**Pull request by arichardson**
This should fix the one failing MIPS cheritest due to missing byval argument copies.
Turns out we were ignoring the byval attribute in the IR, but that rarely mattered since the clang frontend will almost always emit an explicit alloca+mempcy anyway.
For some reason, this is no longer true after the latest merge for -fPIE MIPS code.

### CODE REVIEW

**Comment by jrtc27** [comment]
I guess we get this right in the RISC-V backend because the threshold is 2*XLEN? Or maybe I explicitly made it do the right thing, I forget.

**Comment by arichardson** [comment]
> I guess we get this right in the RISC-V backend because the threshold is 2*XLEN? Or maybe I explicitly made it do the right thing, I forget.

Seems like CC_RISCVAssign2XLen always passes on the stack for RISC-V

**Comment by jrtc27** [comment]
> > I guess we get this right in the RISC-V backend because the threshold is 2*XLEN? Or maybe I explicitly made it do the right thing, I forget.
> 
> Seems like CC_RISCVAssign2XLen always passes on the stack for RISC-V

Oh right yes this is about passing _anything_ on the stack, not forcing capabilities to be on the stack.

File: clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+struct ExecState;
+
+JsValue throwException(ExecState *, JsValue);
+
+// CHECK-LABEL: define {{[^@]+}}@_Z4testv()
+// MIPS: call inreg i8 addrspace(200)* @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)* null, i8 addrspace(200)* inreg %{{[a-z.0-9]+}})
+// RV64: call %union.JsValue @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)* null, i8 addrspace(200)* %{{[a-z.0-9]+}})
+//
+// MIPS: declare inreg i8 addrspace(200)* @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)*, i8 addrspace(200)* inreg)
+// RV64: declare %union.JsValue @_Z14throwExceptionP9ExecState7JsValue(%struct.ExecState addrspace(200)*, i8 addrspace(200)*)
```
Comment (arichardson) [inline clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: @jrtc27 Is it intentional that RISC-V changes the argument type to `i8 addrspace(200)*` but not the return type?

Comment (jrtc27) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: Hm, they're supposed to be handled identically, save for C++ special cases (which shouldn't apply here, and I checked this happens even for C). The only difference _should_ be how many GPRs/FPRs it claims are available (2/{0,2} vs 8/{0,8}). I wonder if this particular conversion is happening somewhere else other than RISCVABIInfo?

Comment (jrtc27) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: Yeah, RISCVABIInfo::classifyArgumentType sees them both as the union and does:
```cxx
  if (IsSingleCapStruct)
    return ABIArgInfo::getDirect();
```

Comment (arichardson) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: The only thing I can see that could break this is
```
  if (!getCXXABI().classifyReturnType(FI))
    FI.getReturnInfo() = classifyReturnType(RetTy);
```
Let's see what the dbugger says.

Comment (jrtc27) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: I wouldn't see `classifyArgumentType` being called twice in that case, and printf debugging says that condition is always true (ie `FI.getReturnInfo() = classifyReturnType(RetTy);` _does_ get called as we expect).

Comment (arichardson) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: Yeah looks like it's happening in CGCall.cpp. I guess for MIPS the ABI info already does the return type flattening.

Comment (arichardson) [reply clang/test/CodeGenCXX/cheri/single-cap-no-byval.cpp:23]: Generate assembly looks sensible so this difference shouldn't matter.

**Comment by jrtc27** [comment]
_sigh_ you didn't do a clean CheriBSD build did you:

```
clang: /local/scratch/jenkins/workspace/CLANG-LLVM-linux_dev/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1261: llvm::SDValue llvm::SelectionDAG::getConstant(const llvm::ConstantInt &, const llvm::SDLoc &, llvm::EVT, bool, bool): Assertion `Int.isNullValue() && "Should not create non-zero capability " "constants with SelectionDAG::getConstant()"' failed.
PLEASE submit a bug report to https://github.com/CTSRD-CHERI/llvm-project/issues and include the crash backtrace, preprocessed source, and associated run script.
If you built clang from source, please provide a reduced test case by running:
	$LLVM_BUILD_DIR/bin/creduce_crash_testcase.py <reproducer>.sh
Stack dump:
0.	Program arguments: /local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang -target riscv64-unknown-freebsd13.0 -fuse-ld=lld -Qunused-arguments -target riscv64-unknown-freebsd13.0 --sysroot=/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd-riscv64-purecap-build/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd/riscv.riscv64c/tmp -B/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd-riscv64-purecap-build/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd/riscv.riscv64c/tmp/usr/bin -O2 -pipe -fno-common -D__LP64__=1 -march=rv64imafdcxcheri -mabi=l64pc128d -fPIE -mno-relax -fPIC -fPIE -MD -MF.depend.display.o -MTdisplay.o -std=gnu99 -Wno-format-zero-length -nobuiltininc -idirafter /local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/lib/clang/11.0.0/include -Wsystem-headers -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wno-error=pass-failed -Wno-error=void-pointer-to-enum-cast -Wno-unknown-warning-option -Wmissing-variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Qunused-arguments -c /local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd/usr.bin/hexdump/display.c -o display.o 
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cheribsd/usr.bin/hexdump/display.c'.
4.	Running pass 'RISCV DAG->DAG Pattern Instruction Selection' on function '@display'
 #0 0x0000000005be8a2d (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5be8a2d)
 #1 0x0000000005be621e llvm::sys::RunSignalHandlers() (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5be621e)
 #2 0x0000000005be7a0d llvm::sys::CleanupOnSignal(unsigned long) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5be7a0d)
 #3 0x0000000005b616e1 (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5b616e1)
 #4 0x0000000005b6197a (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5b6197a)
 #5 0x00007f278cb168a0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x128a0)
 #6 0x00007f278b948f47 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3ef47)
 #7 0x00007f278b94a8b1 abort (/lib/x86_64-linux-gnu/libc.so.6+0x408b1)
 #8 0x00007f278b93a42a (/lib/x86_64-linux-gnu/libc.so.6+0x3042a)
 #9 0x00007f278b93a4a2 (/lib/x86_64-linux-gnu/libc.so.6+0x304a2)
#10 0x0000000006bd5e77 llvm::SelectionDAG::getConstant(llvm::ConstantInt const&, llvm::SDLoc const&, llvm::EVT, bool, bool) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6bd5e77)
#11 0x0000000006bd59f6 llvm::SelectionDAG::getConstant(unsigned long, llvm::SDLoc const&, llvm::EVT, bool, bool) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6bd59f6)
#12 0x0000000006d58c93 llvm::TargetLowering::expandUnalignedLoad(llvm::LoadSDNode*, llvm::SelectionDAG&) const (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6d58c93)
#13 0x0000000006b58b8e (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6b58b8e)
#14 0x0000000006b375eb (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6b375eb)
#15 0x0000000006b36825 llvm::SelectionDAG::Legalize() (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6b36825)
#16 0x0000000006c3cbef llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6c3cbef)
#17 0x0000000006c3a4c5 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6c3a4c5)
#18 0x0000000006c35f2b llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6c35f2b)
#19 0x0000000004f5223e llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x4f5223e)
#20 0x00000000053755ce llvm::FPPassManager::runOnFunction(llvm::Function&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x53755ce)
#21 0x000000000537c5b8 llvm::FPPassManager::runOnModule(llvm::Module&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x537c5b8)
#22 0x0000000005375fd5 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5375fd5)
#23 0x00000000062ca21c clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x62ca21c)
#24 0x00000000066c95b4 (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x66c95b4)
#25 0x0000000006f83084 clang::ParseAST(clang::Sema&, bool, bool) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6f83084)
#26 0x0000000006607582 clang::FrontendAction::Execute() (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6607582)
#27 0x0000000006553832 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x6553832)
#28 0x00000000066c2cc6 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x66c2cc6)
#29 0x00000000038ad776 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x38ad776)
#30 0x00000000038ab4bd (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x38ab4bd)
#31 0x00000000063e4392 (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x63e4392)
#32 0x0000000005b61548 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x5b61548)
#33 0x00000000063e396c clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x63e396c)
#34 0x00000000063a80e9 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x63a80e9)
#35 0x00000000063a8527 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x63a8527)
#36 0x00000000063c517b clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x63c517b)
#37 0x00000000038aa4ea main (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x38aa4ea)
#38 0x00007f278b92bb97 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b97)
#39 0x000000000380190a _start (/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin/clang+0x380190a)
clang-11: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 11.0.0 (https://github.com/CTSRD-CHERI/llvm-project.git b1f774919f4c3d282e2599c6f3f9292fa1ba78d7)
Target: riscv64-unknown-freebsd13.0
Thread model: posix
InstalledDir: /local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-613/cherisdk/bin
clang-11: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-11: note: diagnostic msg: /tmp/display-eff788.c
clang-11: note: diagnostic msg: /tmp/display-eff788.sh
clang-11: note: diagnostic msg: 

********************
--- display.o ---
*** [display.o] Error code 254
```

So that condition does indeed get reached. Probably these two:

```
    SDValue PtrIncrement = DAG.getConstant(RegBytes, dl, PtrVT);
    SDValue StackPtrIncrement = DAG.getConstant(RegBytes, dl, StackPtrVT);
```

There are also a few others in the target-independent parts of the tree, based on a crude 'getConstant(.*PtrVT)' grep, but others likely exist too. I'm going to revert the assertion for now until we've fixed all those.

**Comment by arichardson** [comment]
I did but maybe not RISCV purecap. MIPS definitely worked.

**Pull request merged**